### PR TITLE
dojson: fix arbitrary dojson exceptions

### PR DESCRIPTION
**Pull request by mihaibivol**
Signed-off-by: Mihai Bivol <mihai.bivol@cern.ch>

### CODE REVIEW

**Comment by mihaibivol** [comment]
This yields ```(AttributeError("'tuple' object has no attribute 'lower'",), <type 'exceptions.AttributeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 50, in <module>\n    record = convert_marcxml(StringIO(marcx    ml)).next()\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 59, in convert_marcxml\n    yield strip_empty_values(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py",     line 55, in do\n    add_book_info(output, blob)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 40, in add_book_info\n    collections.append(c.get(\'primary\').lower())\nAttributeError: \'tuple\' obj    ect has no attribute \'lower\'\n')
```

**Comment by mihaibivol** [comment]
This should also be fixed by making `force_list` return a  ing list

**Comment by jacquerie** [comment]
> This should also be fixed by making `force_list` return a  ing list

But you want this merged no matter what, right?

**Comment by mihaibivol** [comment]
@jacquerie Not yet since probably we'd have to do a cleanup of `force_list` returning not lists everywhere else.

**Comment by mihaibivol** [comment]
```     1 (ValueError("invalid literal for int() with base 10: '423-437'",), <type 'exceptions.ValueError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 56, in do\n    add_book_info(output, blob)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 48, in add_book_info\n    int(force_force_list(pubinfo.get(\'0\'))[0]), \'literature\')\nValueError: invalid literal for int() with base 10: \'423-437\'\n'
      1 (ValueError("invalid literal for int() with base 10: '??????a faire'",), <type 'exceptions.ValueError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 56, in do\n    add_book_info(output, blob)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 48, in add_book_info\n    int(force_force_list(pubinfo.get(\'0\'))[0]), \'literature\')\nValueError: invalid literal for int() with base 10: \'??????a faire\'\n'
      4 (IndexError('list index out of range',), <type 'exceptions.IndexError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 53, in overdo_marc_dict\n    return clean_record(conferences.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 65, in wrapper\n    return [f(self, key, values, **kwargs)]\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/conferences/fields/bd1xx.py", line 62, in acronym\n    address = parse_conference_address(raw_address)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/utils/geo.py", line 522, in parse_conference_address\n    us_state = match_us_state(geo_elements[-2].upper().strip()\nIndexError: list index out of range\n'
     12 (AttributeError("'tuple' object has no attribute 'lower'",), <type 'exceptions.AttributeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 55, in do\n    output = super(Publication, self).do(blob, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd01x09x.py", line 95, in persistent_identifiers\n    if val.get("2") and val.get("2", \'\').lower() == "doi":\nAttributeError: \'tuple\' object has no attribute \'lower\'\n'
     12 (ISBNRangeError(), <class 'isbn.hyphen.ISBNRangeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 55, in do\n    output = super(Publication, self).do(blob, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in references\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in <genexpr>\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 80, in get_value\n    method(element)\n  File "/home/mihai/src/inspire-next/inspirehep/modules/references/processors.py", line 242, in add_uid\n    isbn = idutils.normalize_isbn(uid)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/idutils/__init__.py", line 448, in normalize_isbn\n    return ISBN(val).hyphen()\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/isbn.py", line 98, in hyphen\n    return ISBN.hyphenRange.hyphenformat(self._id)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1208, in hyphenformat\n    for i in ISBNRange.hyphensegments(isbn):\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1198, in hyphensegments\n    raise ISBNRangeError(isbn)\nISBNRangeError: \'9799533076521\'\n'
     16 (ISBNRangeError(), <class 'isbn.hyphen.ISBNRangeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 55, in do\n    output = super(Publication, self).do(blob, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in references\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in <genexpr>\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 80, in get_value\n    method(element)\n  File "/home/mihai/src/inspire-next/inspirehep/modules/references/processors.py", line 242, in add_uid\n    isbn = idutils.normalize_isbn(uid)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/idutils/__init__.py", line 448, in normalize_isbn\n    return ISBN(val).hyphen()\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/isbn.py", line 98, in hyphen\n    return ISBN.hyphenRange.hyphenformat(self._id)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1208, in hyphenformat\n    for i in ISBNRange.hyphensegments(isbn):\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1198, in hyphensegments\n    raise ISBNRangeError(isbn)\nISBNRangeError: \'9786611925284\'\n'
     25 (ISBNRangeError(), <class 'isbn.hyphen.ISBNRangeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 55, in do\n    output = super(Publication, self).do(blob, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in references\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in <genexpr>\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 80, in get_value\n    method(element)\n  File "/home/mihai/src/inspire-next/inspirehep/modules/references/processors.py", line 242, in add_uid\n    isbn = idutils.normalize_isbn(uid)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/idutils/__init__.py", line 448, in normalize_isbn\n    return ISBN(val).hyphen()\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/isbn.py", line 98, in hyphen\n    return ISBN.hyphenRange.hyphenformat(self._id)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1208, in hyphenformat\n    for i in ISBNRange.hyphensegments(isbn):\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1198, in hyphensegments\n    raise ISBNRangeError(isbn)\nISBNRangeError: \'9799533073926\'\n'
     25 (TypeError('expected string or buffer',), <type 'exceptions.TypeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 55, in do\n    output = super(Publication, self).do(blob, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 65, in wrapper\n    return [f(self, key, values, **kwargs)]\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 50, in wrapper\n    out = f(*args, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd01x09x.py", line 43, in isbns\n    isbn = normalize_isbn(value[\'a\'])\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/idutils/__init__.py", line 448, in normalize_isbn\n    return ISBN(val).hyphen()\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/isbn.py", line 70, in __init__\n    test = _normalize(str)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/isbn.py", line 23, in _normalize\n    if not re.match(\'^(\\d(-| )?){9}(x|X|\\d|(\\d(-| )?){3}\\d)$\', str):\n  File "/home/mihai/virtualenvs/inspirehep/lib/python2.7/re.py", line 141, in match\n    return _compile(pattern, flags).match(string)\nTypeError: expected string or buffer\n'
     30 (AttributeError("'tuple' object has no attribute 'upper'",), <type 'exceptions.AttributeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 48, in overdo_marc_dict\n    return clean_record(hepnames.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 65, in wrapper\n    return [f(self, key, values, **kwargs)]\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 50, in wrapper\n    out = f(*args, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hepnames/fields/bd1xx.py", line 133, in ids\n    id_type = id_type.upper()\nAttributeError: \'tuple\' object has no attribute \'upper\'\n'
     32 (AttributeError("'tuple' object has no attribute 'upper'",), <type 'exceptions.AttributeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 42, in overdo_marc_dict\n    return clean_record(institutions.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 65, in wrapper\n    return [f(self, key, values, **kwargs)]\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 50, in wrapper\n    out = f(*args, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/institutions/fields/bd1xx.py", line 95, in address\n    force_force_list(value.get(\'g\')),\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/utils/geo.py", line 548, in parse_institution_address\n    state_province = match_us_state(state_province) or state_province\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/utils/geo.py", line 492, in match_us_state\n    state_string = state_string.upper().replace(\'.\', \'\').strip()\nAttributeError: \'tuple\' object has no attribute \'upper\'\n'
     49 (ISBNRangeError(), <class 'isbn.hyphen.ISBNRangeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 55, in do\n    output = super(Publication, self).do(blob, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in references\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 89, in <genexpr>\n    references.extend(get_value(v) for v in value)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/fields/bd90x99x.py", line 80, in get_value\n    method(element)\n  File "/home/mihai/src/inspire-next/inspirehep/modules/references/processors.py", line 242, in add_uid\n    isbn = idutils.normalize_isbn(uid)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/idutils/__init__.py", line 448, in normalize_isbn\n    return ISBN(val).hyphen()\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/isbn.py", line 98, in hyphen\n    return ISBN.hyphenRange.hyphenformat(self._id)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1208, in hyphenformat\n    for i in ISBNRange.hyphensegments(isbn):\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/isbn/hyphen.py", line 1198, in hyphensegments\n    raise ISBNRangeError(isbn)\nISBNRangeError: \'9799533073896\'\n'
    262 (TypeError('float() argument must be a string or a number',), <type 'exceptions.TypeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 42, in overdo_marc_dict\n    return clean_record(institutions.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/schema.py", line 34, in do\n    output = super(SchemaOverdo, self).do(blob, **kwargs)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/overdo.py", line 152, in do\n    data = creator(output, key, value)\n  File "/home/mihai/virtualenvs/inspirehep/local/lib/python2.7/site-packages/dojson/utils.py", line 50, in wrapper\n    out = f(*args, **kwargs)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/institutions/fields/bd1xx.py", line 47, in location\n    \'longitude\': _get_float(value, \'d\'),\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/institutions/fields/bd1xx.py", line 42, in _get_float\n    return float(value[c])\nTypeError: float() argument must be a string or a number\n'
   1049 (AttributeError("'tuple' object has no attribute 'lower'",), <type 'exceptions.AttributeError'>, 'Traceback (most recent call last):\n  File "./dojsn.py", line 56, in <module>\n    keep_singletons=False))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/processors.py", line 55, in overdo_marc_dict\n    return clean_record(hep.do(record))\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 56, in do\n    add_book_info(output, blob)\n  File "/home/mihai/src/inspire-next/inspirehep/dojson/hep/model.py", line 41, in add_book_info\n    collections.append(c.get(\'primary\').lower())\nAttributeError: \'tuple\' object has no attribute \'lower\'\n'
   2047 
```

**Comment by mihaibivol** [comment]
@jacquerie after the build passes this is ready to review.

This PR touches most of the latest errors found when migrating production. Some of the errors might be fixed directly there (such as normalizing dashes). These should be carefully monitored as part of #1365

File: inspirehep/dojson/hep/fields/bd01x09x.py [hunk 1]
```diff
@@ -40,8 +42,11 @@
 def isbns(self, key, value):
     "ISBN, its medium and an additional comment."""
     try:
-        isbn = normalize_isbn(value['a'])
-    except (KeyError, ISBNError):
+        isbn = normalize_isbn(force_single_element(value['a']))
+    # See https://github.com/nekobcn/isbnid/issues/2 and
+    # https://github.com/nekobcn/isbnid/issues/3 for understanding the long
+    # exception list.
```
Comment (jacquerie) [inline inspirehep/dojson/hep/fields/bd01x09x.py]: Makes me wonder if we should handle this complexity in `idutils`.

**Comment by jacquerie** [comment]
LGTM :ship:

**Pull request merged**